node/core/provisioner/src/lib.rs

```diff
@@ -551,56 +555,135 @@ fn bitfields_indicate_availability(
 }
 
-async fn select_disputes(
+/// Maximum number of candidates we query at once for dispute statements.
+const MAX_CHUNK_SIZE: usize = 100;
+
+#[derive(Debug)]
+enum RequestType {
+	/// Query recent disputes, could be an excessive amount.
+	Recent,
+	/// Query the currently active and very recently concluded disputes.
+	Active,
+}
+
+/// Request open disputes identified by `CandidateHash` and the `SessionIndex`.
+async fn request_disputes(
 	sender: &mut impl SubsystemSender,
-	metrics: &metrics::Metrics,
-) -> Result<MultiDisputeStatementSet, Error> {
+	active_or_recent: RequestType,
+) -> Vec<(SessionIndex, CandidateHash)> {
 	let (tx, rx) = oneshot::channel();
-
-	// We use `RecentDisputes` instead of `ActiveDisputes` because redundancy is fine.
-	// It's heavier than `ActiveDisputes` but ensures that everything from the dispute
-	// window gets on-chain, unlike `ActiveDisputes`.
-	//
-	// This should have no meaningful impact on performance on production networks for
-	// two reasons:
-	// 1. In large validator sets, a node might be a block author 1% or less of the time.
-	//    this code-path only triggers in the case of being a block author.
-	// 2. Disputes are expected to be rare because they come with heavy slashing.
-	sender.send_message(DisputeCoordinatorMessage::RecentDisputes(tx).into()).await;
+	let msg = match active_or_recent {
+		RequestType::Recent => DisputeCoordinatorMessage::RecentDisputes(tx),
+		RequestType::Active => DisputeCoordinatorMessage::ActiveDisputes(tx),
+	};
+	sender.send_message(msg.into()).await;
 
 	let recent_disputes = match rx.await {
 		Ok(r) => r,
 		Err(oneshot::Canceled) => {
-			tracing::debug!(
-				target: LOG_TARGET,
-				"Unable to gather recent disputes - subsystem disconnected?",
-			);
-
+			tracing::debug!(target: LOG_TARGET, "Unable to gather {:?} disputes", active_or_recent,);
 			Vec::new()
 		},
 	};
+	recent_disputes
+}
 
-	// Load all votes for all disputes from the coordinator.
-	let dispute_candidate_votes = {
-		let (tx, rx) = oneshot::channel();
-		sender
-			.send_message(
-				DisputeCoordinatorMessage::QueryCandidateVotes(recent_disputes, tx).into(),
-			)
-			.await;
+/// Request the relevant dispute statements for a set of disputes identified by `CandidateHash` and the `SessionIndex`.
+async fn request_votes(
+	sender: &mut impl SubsystemSender,
+	disputes_to_query: Vec<(SessionIndex, CandidateHash)>,
+) -> Vec<(SessionIndex, CandidateHash, CandidateVotes)> {
+	let (tx, rx) = oneshot::channel();
+	sender
+		.send_message(DisputeCoordinatorMessage::QueryCandidateVotes(disputes_to_query, tx).into())
+		.await;
 
-		match rx.await {
-			Ok(v) => v,
-			Err(oneshot::Canceled) => {
-				tracing::debug!(
-					target: LOG_TARGET,
-					"Unable to query candidate votes - subsystem disconnected?",
-				);
-				Vec::new()
-			},
+	match rx.await {
+		Ok(v) => v,
+		Err(oneshot::Canceled) => {
+			tracing::debug!(target: LOG_TARGET, "Unable to query candidate votes",);
```

Comment:
```suggestion
			tracing::debug!(target: LOG_TARGET, "Unable to query candidate votes");
```

Reply:
I am not familiar with the logic here, but isn't this worth a higher log level?

Reply:
I kept it at the same level, warn might indeed be the better approach.